Added mcpKadjp function.

bwettimz opened the pull request:
Added basic part of mcpJadjp function in utilities.py. Only Holm's method is available, but I will add the other methods in next push.

bwettimz commented on hypothesize/utilities.py:
```diff
@@ -2261,6 +2261,33 @@ def linhat(x, con, est, *args):
+def mcpKadjp(p,proc="Holm"):
+    
+    """
+    P-value adjustment procedure based on Holms method..
```
Need to flesh out.

Alcampopiano replied:
I really appreciate this by the way. I will look things over as soon as I get a spare second at work.

bwettimz replied:
No worries. Sorry for not adding the requested functions. Only saw the list after the function had been created for a different project im working on. Will slowly add some from the list.

Alcampopiano replied:
A couple notes having looked things over:

- The results in Wilcox's mcpKadjp package do not seem to match the results in your function. I tried with `p=[0.010 0.030 0.011 0.003]` in both R and Python and got slightly different results.

- To make the output easier to read, the output DataFrame should have labelled columns that match what Wilcox uses. For example,
![scw](https://user-images.githubusercontent.com/4624246/95932215-3c6f0c00-0d99-11eb-8958-0171edf04396.png)

- It would be great to have a doc string that is close to what Wilcox uses in his WRS package (e.g., with the citations). It would also be nice if the doc string matched the format used in other user-facing functions (for example, see [`con1way`](https://github.com/Alcampopiano/hypothesize/blob/56d75ade8cd2341542a770d05ebead3eef45d363/hypothesize/utilities.py#L10) or any of the hypothesis testing functions)

If these issues can be addressed, I would be happy to accept the PR. I would likely even add another correction method so that the function is set up to handle a list of correction methods (as it is in R).

bwettimz commented:
**Changes:**
- Function should work now as intended. 
- Also added Hochberg method.
- Updated doc string

**PYTHON OUTPUT**
     rawp    Holm
0  0.7800  1.0000
1  0.0100  0.0600
2  0.0300  0.1200
3  0.0110  0.0600
4  0.0030  0.0210
5  0.5600  1.0000
6  0.1200  0.3600
7  0.0001  0.0008

     rawp  Hochberg
0  0.7800    0.7800
1  0.0100    0.0550
2  0.0300    0.1200
3  0.0110    0.0550
4  0.0030    0.0210
5  0.5600    0.7800
6  0.1200    0.3600
7  0.0001    0.0008

**R OUTPUT**
       rawp   Holm
[1,] 0.7800 1.0000
[2,] 0.0100 0.0600
[3,] 0.0300 0.1200
[4,] 0.0110 0.0600
[5,] 0.0030 0.0210
[6,] 0.5600 1.0000
[7,] 0.1200 0.3600
[8,] 0.0001 0.0008

       rawp Hochberg
[1,] 0.7800   0.7800
[2,] 0.0100   0.0550
[3,] 0.0300   0.1200
[4,] 0.0110   0.0550
[5,] 0.0030   0.0210
[6,] 0.5600   0.7800
[7,] 0.1200   0.3600
[8,] 0.0001   0.0008

Alcampopiano commented on hypothesize/utilities.py:
```diff
@@ -3,6 +3,7 @@
 from scipy.stats import trim_mean
 from scipy.stats import t
 import pandas as pd
+import sys
```
We could avoid this import since it is only used here. The error that is raised in your function could instead use `raise Exception("error message here"). What do you think?

bwettimz replied:
I agree! Think it was just out of habit haha!

Alcampopiano commented on hypothesize/utilities.py:
```diff
@@ -2261,32 +2262,69 @@ def linhat(x, con, est, *args):
 
-def mcpKadjp(p,proc="Holm"):
+def mcpKadjp(p, k=1, proc = 'Holm'):
+    '''
```
These should be triple _double_ quotes. For example,

````
"""
some string
"""

bwettimz replied:
Changed as per your suggestion.

Alcampopiano commented on hypothesize/utilities.py:
```diff
@@ -2261,32 +2262,69 @@ def linhat(x, con, est, *args):
+    The  value for k-FWER. Defaults at to 1.
+    
+    :param proc: str
+    Indicates the method to be used. Choices are: ' Holm' ,'Hochberg'.
```
Space in the Holm string should be removed

bwettimz replied:
Changed as per your suggestion.

Alcampopiano commented on hypothesize/utilities.py:
```diff
@@ -2261,32 +2262,69 @@ def linhat(x, con, est, *args):
-    return pd.DataFrame(adjp)
+    adjp = pd.DataFrame(adjp)
+    adjp.columns = ['rawp', f'{proc}']
+    print(adjp)
```
There should be a return statement instead of a print statement

bwettimz replied:
Changed as per your suggestion.

Alcampopiano commented on hypothesize/utilities.py:
```diff
@@ -2261,32 +2262,69 @@ def linhat(x, con, est, *args):
     adjp[:,1] = [tmp[s] for s in index_list]
-    return pd.DataFrame(adjp)
+    adjp = pd.DataFrame(adjp)
+    adjp.columns = ['rawp', f'{proc}']
```
Is the f string necessary since the only variable used is the proc string itself?

bwettimz replied:
I agree!

Alcampopiano commented:
@bwettimz 

Hello,

Unfortunately, given the same input, Wilcox's R function still produces different values when compared to your function. For example,

In Python:

```python
p=[0.01, 0.05, 0.02]
mcpKadjp(, k=.4, proc = 'Holm')
```

   rawp  Holm
0  0.01  0.03
1  0.05  0.05
2  0.02  0.04

In R:

```R
p=c(.01, .05, .02)
mcpKadjp(p, k=.4, proc='Holm', rawp=p)
```
rawp  Holm
[1,] 0.01 0.075
[2,] 0.05 0.125
[3,] 0.02 0.100

The formulas for `crit` appear to be incorrect in the Python code from what I can tell.

Alcampopiano commented:
@bwettimz I'm kindly closing this PR due to inactivity. I'm happy to reopen it in the future. Take care.
